cedar-policy-core/src/ast/policy.rs

```diff
@@ -960,91 +1009,221 @@ pub struct TemplateBody {
     non_scope_constraints: Arc<Expr>,
 }
 
+#[derive(Clone, Hash, Eq, PartialEq, Debug)]
+/// Policy datatype. This is used for both templates (in which case it contains
+/// slots) and static policies (in which case it contains zero slots).
+pub enum TemplateBody {
+    /// Represents a valid template body
+    TemplateBody(TemplateBodyImpl),
+    #[cfg(feature = "tolerant-ast")]
+    /// Represents a policy that failed to parse
+    TemplateBodyError(PolicyID),
+}
+
+impl Serialize for TemplateBody {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            TemplateBody::TemplateBody(impl_body) => impl_body.serialize(serializer),
+            #[cfg(feature = "tolerant-ast")]
+            TemplateBody::TemplateBodyError(policy_id) => {
+                format!("TemplateBody::TemplateBodyError({policy_id})").serialize(serializer)
+            }
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for TemplateBody {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let impl_body = TemplateBodyImpl::deserialize(deserializer)?;
+        Ok(TemplateBody::TemplateBody(impl_body))
+    }
+}
+
 impl TemplateBody {
     /// Get the `Id` of this policy.
     pub fn id(&self) -> &PolicyID {
-        &self.id
+        match self {
+            TemplateBody::TemplateBody(TemplateBodyImpl { id, .. }) => &id,
+            #[cfg(feature = "tolerant-ast")]
+            TemplateBody::TemplateBodyError(policy_id) => &policy_id,
+        }
     }
 
     /// Get the location of this policy
     pub fn loc(&self) -> Option<&Loc> {
-        self.loc.as_ref()
+        match self {
+            TemplateBody::TemplateBody(TemplateBodyImpl { loc, .. }) => loc.as_ref(),
+            #[cfg(feature = "tolerant-ast")]
+            TemplateBody::TemplateBodyError(_) => None,
+        }
     }
 
     /// Clone this policy with a new `Id`.
     pub fn new_id(&self, id: PolicyID) -> Self {
-        let mut new = self.clone();
+        let existing_template_body = match self {
+            TemplateBody::TemplateBody(t) => t,
+            #[cfg(feature = "tolerant-ast")]
+            TemplateBody::TemplateBodyError(_) => return TemplateBody::TemplateBodyError(id),
```

Comment:
I'd like to avoid the mix of `match` and `return` here. We could either check first if it's an error and use the `return` (leaving everything else the same), or extend the `match` to cover the entire body of this method.

Reply:
Good point, will do